alerting/src/main/kotlin/org/opensearch/alerting/MonitorRunner.kt

```diff
@@ -757,39 +760,133 @@ object MonitorRunner : JobRunner, CoroutineScope, AbstractLifecycleComponent() {
             return
         }
 
-        for (query in queries) {
-            runForEachQuery(monitor, lastRunContext, index, query)
+        val count: Int = lastRunContext["shards_count"] as Int
+        val updatedLastRunContext = lastRunContext.toMutableMap()
+        for (i: Int in 0 until count) {
+            val shard = i.toString()
+            val maxSeqNo: Long = getMaxSeqNo(index, shard)
+            updatedLastRunContext[shard] = maxSeqNo.toString()
+        }
+
+        val queryToDocIds = mutableMapOf<DocLevelQuery, Set<String>>()
+        val docsToQueries = mutableMapOf<String, MutableList<String>>()
+        val docExecutionContext = DocumentExecutionContext(queries, lastRunContext, updatedLastRunContext)
+        queries.forEach { query ->
+            val matchingDocIds = runForEachQuery(docExecutionContext, query, index)
+            queryToDocIds[query] = matchingDocIds
+            matchingDocIds.forEach {
+                docsToQueries.putIfAbsent(it, mutableListOf())
+                docsToQueries[it]?.add(query.id)
+            }
         }
+
+        val queryIds = queries.map { it.id }
+
+        monitor.triggers.forEach {
+            runForEachDocTrigger(it as DocumentLevelTrigger, monitor, docsToQueries, queryIds, queryToDocIds)
+        }
+
+        // TODO: Check for race condition against the update monitor api
+        // This does the update at the end in case of errors and makes sure all the queries are executed
+        val updatedMonitor = monitor.copy(lastRunContext = updatedLastRunContext)
+        // note: update has to called in serial for shards of a given index.
+        // make sure this is just updated for the specific query or at the end of all the queries
+        updateMonitor(client, xContentRegistry, settings, updatedMonitor)
```

Comment:
I will remove these comments since these have now been sorted. Yea there has been talks of having another index for additional metadata information and this can help with throttling.